### PR TITLE
refactor: make unleash client testable

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Making Unleash client testable by using dependency injection.

In a followup PR I will write actual tests for metrics since we have this option now.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -62,50 +61,14 @@ class Client extends EventEmitter implements IClient {
 
     private ready: boolean = false;
 
-    constructor(config: IProxyConfig, init: Function = initialize) {
+    constructor(config: IProxyConfig, unleash: Unleash, metrics: Metrics) {
```
Comment (kwasniew) [inline src/client.ts:64]: dependencies are injected instead of client building them internally

File: src/create-client.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { IProxyConfig } from './config';
+import Client from './client';
+import { initialize } from 'unleash-client';
+import Metrics from 'unleash-client/lib/metrics';
+import { defaultStrategies } from 'unleash-client/lib/strategy';
+
+export const createClient = (config: IProxyConfig): Client => {
```
Comment (kwasniew) [inline src/create-client.ts:7]: we moved the actual assembly of dependencies to this new file that owns this responsibility making client much simpler and more testable

File: src/test/client.test.ts [hunk 3]
```diff
@@ -21,7 +14,7 @@ test('should add environment to isEnabled calls', () => {
 
     config.disableMetrics = true;
 
-    const client = new Client(config, init);
+    const client = createFakeClient(config);
```
Comment (kwasniew) [inline src/test/client.test.ts:17]: we have another assembly function for testing purposes

File: src/client.ts [hunk 4]
```diff
@@ -62,50 +61,14 @@ class Client extends EventEmitter implements IClient {
...
-            customHeadersFunction,
-            bootstrap: config.bootstrap,
-            storageProvider: config.storageProvider,
-            ...(!!config.httpOptions
-                ? { httpOptions: config.httpOptions }
-                : {}),
-        });
-
-        // Custom metrics Instance
-        this.metrics = new Metrics({
```
Comment (kwasniew) [inline src/client.ts:96]: it was impossible to test because of this hardcoded dependency

File: src/client.ts [hunk 5]
```diff
@@ -62,50 +61,14 @@ class Client extends EventEmitter implements IClient {
 
     private ready: boolean = false;
 
-    constructor(config: IProxyConfig, init: Function = initialize) {
```
Comment (kwasniew) [inline src/client.ts:65]: I'm not a fan of the partial injection inside a constructor. It should be a responsibility of a composition root (in our case it's create-client.ts)

File: src/create-client.ts (hunk 2 above)
Comment (kwasniew) [reply src/create-client.ts:7]: This is so-called composition root: https://blog.ploeh.dk/2011/07/28/CompositionRoot/

**Review by Tymek — APPROVED** [review]
Nice refactor. Thank you for clear explanation

**Pull request merged**